src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/transform/TransformMetadataService.kt

```diff
@@ -0,0 +1,119 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.indexmanagement.transform
+
+import com.amazon.opendistroforelasticsearch.indexmanagement.IndexManagementPlugin
+import com.amazon.opendistroforelasticsearch.indexmanagement.elasticapi.parseWithType
+import com.amazon.opendistroforelasticsearch.indexmanagement.elasticapi.suspendUntil
+import com.amazon.opendistroforelasticsearch.indexmanagement.transform.exceptions.TransformMetadataException
+import com.amazon.opendistroforelasticsearch.indexmanagement.transform.model.Transform
+import com.amazon.opendistroforelasticsearch.indexmanagement.transform.model.TransformMetadata
+import com.amazon.opendistroforelasticsearch.indexmanagement.transform.model.TransformStats
+import com.amazon.opendistroforelasticsearch.indexmanagement.util.IndexUtils.Companion.hashToFixedSize
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.action.DocWriteRequest
+import org.elasticsearch.action.get.GetRequest
+import org.elasticsearch.action.get.GetResponse
+import org.elasticsearch.action.index.IndexRequest
+import org.elasticsearch.action.index.IndexResponse
+import org.elasticsearch.client.Client
+import org.elasticsearch.common.xcontent.LoggingDeprecationHandler
+import org.elasticsearch.common.xcontent.NamedXContentRegistry
+import org.elasticsearch.common.xcontent.XContentFactory
+import org.elasticsearch.common.xcontent.XContentHelper
+import org.elasticsearch.common.xcontent.XContentType
+import java.time.Instant
+import org.elasticsearch.ExceptionsHelper
+import org.elasticsearch.action.DocWriteResponse
+import org.elasticsearch.common.xcontent.ToXContent
+import org.elasticsearch.transport.RemoteTransportException
+
+@SuppressWarnings("ReturnCount")
+class TransformMetadataService(private val esClient: Client, val xContentRegistry: NamedXContentRegistry) {
+
+    private val logger = LogManager.getLogger(javaClass)
+
+    @Suppress("BlockingMethodInNonBlockingContext")
+    suspend fun getMetadata(transform: Transform): TransformMetadata {
+        return if (transform.metadataId != null) {
+            // update metadata
+            val getRequest = GetRequest(IndexManagementPlugin.INDEX_MANAGEMENT_INDEX, transform.metadataId).routing(transform.id)
+            val response: GetResponse = esClient.suspendUntil { get(getRequest, it) }
+            val metadataSource = response.sourceAsBytesRef
+            val transformMetadata = metadataSource?.let {
+                withContext(Dispatchers.IO) {
+                    val xcp = XContentHelper.createParser(xContentRegistry, LoggingDeprecationHandler.INSTANCE, metadataSource, XContentType.JSON)
+                    xcp.parseWithType(response.id, response.seqNo, response.primaryTerm, TransformMetadata.Companion::parse)
```

Comment:
This can throw exception, did you need it handled in this PR or it will be done in later one?

Reply:
Yeah, there are still some edge case handling I haven't finalized, exception handling and message population for those exceptions. Will take care of this in another PR